besu/src/main/java/org/hyperledger/besu/cli/subcommands/storage/TrieLogHelper.java

```diff
@@ -0,0 +1,227 @@
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.worldstate.DataStorageConfiguration;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.common.base.Splitter;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.bouncycastle.util.encoders.Base64;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Helper class for counting and pruning trie logs */
+public class TrieLogHelper {
+  private static final String trieLogFile = "trieLogsToRetain.txt";
+  private static final Logger LOG = LoggerFactory.getLogger(TrieLogHelper.class);
+
+  static void prune(
+      final PrintWriter out,
+      final DataStorageConfiguration config,
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final MutableBlockchain blockchain) throws IOException {
+
+    TrieLogHelper.validatePruneConfiguration(config);
+    final long layersToRetain = config.getUnstable().getBonsaiTrieLogRetentionThreshold();
+    final long chainHeight = blockchain.getChainHeadBlockNumber();
+    final long lastBlockToRetainTrieLogsFor = chainHeight - layersToRetain;
+    final Optional<Hash> finalizedBlockHash = blockchain.getFinalized();
+    if (finalizedBlockHash.isEmpty()) {
+      LOG.error("No finalized block present, skipping pruning");
+      return;
+    } else {
+      if (blockchain.getBlockHeader(finalizedBlockHash.get()).get().getNumber()
+          < lastBlockToRetainTrieLogsFor) {
+        LOG.error("Trying to prune more layers than the finalized block height, skipping pruning");
+        return;
+      }
+    }
+
+    // retrieve the layersToRetains hashes from blockchain
+    final List<Hash> hashesToRetain = new ArrayList<>();
+
+    for (long i = chainHeight; i > lastBlockToRetainTrieLogsFor; i--) {
+      final Optional<BlockHeader> header = blockchain.getBlockHeader(i);
+      header.ifPresent(blockHeader -> hashesToRetain.add(blockHeader.getHash()));
+    }
+
+    IdentityHashMap<byte[], byte[]> trieLogsToRetain;
+    if ((long) hashesToRetain.size() == layersToRetain) {
+      trieLogsToRetain = new IdentityHashMap<>();
+      // save trielogs in a flatfile as a fail-safe
+      out.println("Obtaining trielogs to retain...");
+      hashesToRetain.forEach(
+          hash -> {
+            rootWorldStateStorage
+                .getTrieLog(hash)
+                .ifPresent(trieLog -> trieLogsToRetain.put(hash.toArrayUnsafe(), trieLog));
+          });
+      out.println("Saving trielogs to retain in file...");
+      saveTrieLogsInFile(trieLogsToRetain);
```

Comment:
Not sure how the fail-safe is being used. Expected to see that if we got an error clearing and inserting entries we ready from the file or allow specifying an existing backup file to restore from.

Reply:
Added a comment and instructions that will be logged for the user. Basically the idea here is, we retain the logs in a file in case something goes wrong.
In that case the user can re-run the subcommand and we will read the trielogs from the file since the column family might have been dropped already.